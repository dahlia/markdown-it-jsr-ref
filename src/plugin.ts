import { escape } from "@std/html";
import type MarkdownIt from "markdown-it";
import type { PluginSimple } from "markdown-it";
import {
  fetchIndex,
  fetchLatestVersion,
  type Index,
  type PackageName,
  type Version,
} from "./jsr.ts";

/**
 * Options for the {@link jsrRef} plugin.
 */
export interface Options {
  package: PackageName;
  version?: Version | "stable" | "unstable";
  cachePath?: string;
}

/**
 * Create a markdown-it plugin to render links to JSR references.
 * @param options Options for the plugin.
 * @returns A markdown-it plugin.
 */
export async function jsrRef(options: Options): Promise<PluginSimple> {
  let version: Version | null;
  if (options.version == null || options.version === "stable") {
    version = await fetchLatestVersion(options.package);
  } else if (options.version === "unstable") {
    version = await fetchLatestVersion(options.package, true);
  } else {
    version = options.version;
  }
  if (version == null) throw new Error("No version found.");
  let cachedIndex: Index | null = null;
  if (options.cachePath != null) {
    // deno-lint-ignore no-explicit-any
    let cacheJson: any | null = null;
    try {
      const cache = await Deno.readTextFile(options.cachePath);
      cacheJson = JSON.parse(cache);
    } catch (_) {
      // Ignore errors.
    }
    if (
      cacheJson != null && cacheJson.package === options.package &&
      cacheJson.version === version
    ) {
      cachedIndex = cacheJson.index;
    }
  }
  const index = cachedIndex ?? await fetchIndex(
    options.package,
    version,
  );
  if (options.cachePath != null && cachedIndex == null) {
    await Deno.writeTextFile(
      options.cachePath,
      JSON.stringify({
        package: options.package,
        version,
        index,
      }),
    );
  }
  return (md: MarkdownIt) => {
    const originalHook = md.renderer.rules.code_inline;
    md.renderer.rules.code_inline = (
      // deno-lint-ignore no-explicit-any
      tokens: any,
      // deno-lint-ignore no-explicit-any
      idx: any,
      // deno-lint-ignore no-explicit-any
      options: any,
      // deno-lint-ignore no-explicit-any
      env: any,
      // deno-lint-ignore no-explicit-any
      self: any,
    ) => {
      let depth = 0;
      for (let i = idx - 1; i >= 0; i--) {
        const token = tokens[i];
        if (token.type === "link_close") {
          depth++;
        } else if (token.type === "link_open") {
          depth--;
          if (depth < 0) break;
        }
      }
      const token = tokens[idx];
      let label = token.content;
      const entry = depth < 0 ? null : index[label];
      if (entry != null) {
        label = entry.label;
      }
      const code = `<code${self.renderAttrs(token)}>${escape(label)}</code>`;
      if (entry == null) {
        if (originalHook) {
          return originalHook(tokens, idx, options, env, self);
        }
        return code;
      }
      return `<a href="${escape(entry.url)}">${code}</a>`;
    };
  };
}
