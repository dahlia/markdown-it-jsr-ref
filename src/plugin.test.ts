import { assertEquals } from "@std/assert/equals";
import { assertSnapshot } from "@std/testing/snapshot";
import MarkdownIt from "markdown-it";
import { jsrRef } from "./plugin.ts";

const text = `\
- \`foo\`
- \`Range\`
- \`new Range()\`
- \`Range.at()\`
- \`~Range.at()\`
- \`take()\`
- [code already inside a link: \`Range\`](test.html)
`;

const text2 = `
- \`foo\`
- \`Federation\`
- \`new MemoryKvStore()\`
- \`Federation.fetch()\`
- \`configure()\`
- \`Logger\`
- \`Logger.with()\`
`;

Deno.test("jsrRef()", async (t) => {
  const md = new MarkdownIt();
  md.use(
    await jsrRef({
      package: "@hongminhee/aitertools",
      version: "0.6.0",
    }),
  );
  const rendered = md.render(text);
  await assertSnapshot(t, rendered);

  const md2 = new MarkdownIt();
  md2.use(
    await jsrRef({
      package: "@hongminhee/aitertools",
      version: "0.6.0",
      cachePath: ".jsr-cache.json",
    }),
  );
  const rendered2 = md.render(text);
  assertEquals(rendered, rendered2);

  const md3 = new MarkdownIt();
  md3.use(
    await jsrRef({
      package: "@fedify/fedify",
      version: "1.4.5",
      cachePath: ".jsr-cache2.json",
    }),
  );
  md3.use(
    await jsrRef({
      package: "@logtape/logtape",
      version: "0.8.2",
      cachePath: ".jsr-cache3.json",
    }),
  );
  const rendered3 = md3.render(text2);
  await assertSnapshot(t, rendered3);
});

// cSpell: ignore hongminhee aitertools
