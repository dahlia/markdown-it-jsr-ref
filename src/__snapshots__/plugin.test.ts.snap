export const snapshot = {};

snapshot[`jsrRef() 1`] = `
'<ul>
<li><code>foo</code></li>
<li><a href="https://jsr.io/@hongminhee/aitertools@0.6.0/doc/~/Range"><code>Range</code></a></li>
<li><a href="https://jsr.io/@hongminhee/aitertools@0.6.0/doc/~/Range#constructor_0"><code>new Range()</code></a></li>
<li><a href="https://jsr.io/@hongminhee/aitertools@0.6.0/doc/~/Range.prototype.at"><code>Range.at()</code></a></li>
<li><a href="https://jsr.io/@hongminhee/aitertools@0.6.0/doc/~/Range.prototype.at"><code>at()</code></a></li>
<li><a href="https://jsr.io/@hongminhee/aitertools@0.6.0/doc/~/take"><code>take()</code></a></li>
<li><a href="test.html">code already inside a link: <code>Range</code></a></li>
</ul>
'
`;

snapshot[`jsrRef() 2`] = `
'<ul>
<li><code>foo</code></li>
<li><a href="https://jsr.io/@fedify/fedify@0.11.3/doc/federation/~/Federation"><code>Federation</code></a></li>
<li><a href="https://jsr.io/@fedify/fedify@0.11.3/doc/federation/~/Federation#constructor_0"><code>new Federation()</code></a></li>
<li><a href="https://jsr.io/@fedify/fedify@0.11.3/doc/federation/~/Federation.prototype.fetch"><code>Federation.fetch()</code></a></li>
</ul>
'
`;
